### ISSUE TITLE
v0.5.0 Site Controller Crashes on Boot

### ISSUE THREAD

**Issue by Kampe** [skupperproject/skupper#442]
Decided to attempt to roll the latest skupper version out (v0.5.0) via the site controller to a cluster that had been running v0.4.2.

Once the `site-controller` starts, these are the logs we obvserve. 

```
2021/03/11 21:33:56 Skupper site controller watching all namespaces
2021/03/11 21:33:56 Starting the Skupper site controller informers
2021/03/11 21:33:56 Waiting for informer caches to sync
2021/03/11 21:33:56 Checking if sites need updates (0.5.0)
E0311 21:33:56.673919       1 runtime.go:78] Observed a panic: "invalid memory address or nil pointer dereference" (runtime error: invalid memory address or nil pointer dereference)
goroutine 1 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x1321120, 0x200cdf0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/util/runtime/runtime.go:48 +0x82
panic(0x1321120, 0x200cdf0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/skupperproject/skupper/client.(*VanClient).RouterUpdateVersionInNamespace(0xc000289980, 0x168c5a0, 0xc000046230, 0x0, 0xc0000473e9, 0x7, 0xc0000cfdd0, 0xc0000b4000, 0x26)
	/go/src/app/client/router_update.go:133 +0x590
main.(*SiteController).updateChecks(0xc00031d450)
	/go/src/app/cmd/site-controller/controller.go:437 +0xff
main.(*SiteController).Run(0xc00031d450, 0xc000154420, 0x0, 0x0)
	/go/src/app/cmd/site-controller/controller.go:129 +0x36c
main.main()
	/go/src/app/cmd/site-controller/main.go:55 +0x117
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x148 pc=0x11ce630]

goroutine 1 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/util/runtime/runtime.go:55 +0x105
panic(0x1321120, 0x200cdf0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/skupperproject/skupper/client.(*VanClient).RouterUpdateVersionInNamespace(0xc000289980, 0x168c5a0, 0xc000046230, 0x0, 0xc0000473e9, 0x7, 0xc0000cfdd0, 0xc0000b4000, 0x26)
	/go/src/app/client/router_update.go:133 +0x590
main.(*SiteController).updateChecks(0xc00031d450)
	/go/src/app/cmd/site-controller/controller.go:437 +0xff
main.(*SiteController).Run(0xc00031d450, 0xc000154420, 0x0, 0x0)
	/go/src/app/cmd/site-controller/controller.go:129 +0x36c
main.main()
	/go/src/app/cmd/site-controller/main.go:55 +0x117
```

We configure using typical YAML configuration.

```
apiVersion: v1
kind: Secret
metadata:
  name: skupper-site-token
  namespace: default
  annotations:
    argocd.argoproj.io/sync-options: Prune=false
    argocd.argoproj.io/compare-options: IgnoreExtraneous
  labels:
    skupper.io/type: connection-token-request
    
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: skupper-site
  namespace: default
  annotations:
    sidecar.istio.io/inject: "false"
data:
  cluster-local: "true"
  console: "true"
  console-authentication: internal
  console-user: "admin"
  console-password: "password"
  edge: "false"
  name: product
  router-console: "true"
  service-controller: "true"
  service-sync: "true"
```

**Comment by grs** [comment]
Sorry! I think I see the problem and will fix.

In the meantime can you check whether you have a service called 'skupper' in the site namespaces, delete it if so and restart the site-controller? (or else revert to previous version of site controller until I get a fix out)

**Comment by Kampe** [comment]
That's quite alright!

So we have two namespaces being skuppered, I deleted both `skupper` services in them and restarted the `site-controller` and still unfortunately see the same behavior.

**Comment by grs** [comment]
I've re-released 0.5.0 with this fix. One thing to point out is that the role the site site controller runs under will need to be updated to allow it access (get, list, watch) to DaemonSets, so it can grant that permission to the service controller.

Apologies again for the glitch and thanks for bringing to our attention!

**Comment by Kampe** [comment]
Excellent! Will update the role and give it a shot!

Thanks!

**Comment by Kampe** [comment]
So our policy currently looks like:

```
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  labels:
    application: skupper-site-controller
  name: skupper-site-controller
rules:
- apiGroups:
  - ""
  resources:
  - configmaps
  - pods
  - pods/exec
  - services
  - secrets
  - serviceaccounts
  verbs:
  - get
  - list
  - watch
  - create
  - update
  - delete
- apiGroups:
  - apps
  resources:
  - deployments
  - statefulsets
  - daemonsets
  verbs:
  - get
  - list
  - watch
  - create
  - update
  - delete
- apiGroups:
  - route.openshift.io
  resources:
  - routes
  verbs:
  - get
  - list
  - watch
  - create
  - delete
- apiGroups:
  - rbac.authorization.k8s.io
  resources:
  - rolebindings
  - roles
  verbs:
  - get
  - list
  - watch
  - create
  - delete
---
```

Haven't rolled this out yet, but as far as I can tell this should be correct though? Does the site controller role still need all of these permissions as well?

**Comment by grs** [comment]
Yes, that looks correct. The site controller needs them only because it needs to give those premissions to the service-controller when it starts. (We do have plans to allow the service account used by skupper to be supplied as a parameter, meaning it could be created by some other means, which would avoid the need for the site controller to have so wide a role).

**Comment by Kampe** [comment]
Unfortunately still see the same issues (or similar) with 0.5.2:
```
2021/03/31 16:22:06 Waiting for informer caches to sync
2021/03/31 16:22:06 Checking if sites need updates (0.5.2)
E0331 16:22:07.179572       1 runtime.go:78] Observed a panic: "assignment to entry in nil map" (assignment to entry in nil map)
goroutine 1 [running]:
k8s.io/apimachinery/pkg/util/runtime.logPanic(0x13225a0, 0x16466a0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/util/runtime/runtime.go:74 +0xa3
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/util/runtime/runtime.go:48 +0x82
panic(0x13225a0, 0x16466a0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/skupperproject/skupper/client.(*VanClient).RouterUpdateVersionInNamespace(0xc0002a2000, 0x168e720, 0xc000048230, 0x0, 0xc000392900, 0xa, 0xc0000c3dd0, 0xc0000a8000, 0x26)
	/go/src/app/client/router_update.go:140 +0x3a10
main.(*SiteController).updateChecks(0xc000205720)
	/go/src/app/cmd/site-controller/controller.go:444 +0xff
main.(*SiteController).Run(0xc000205720, 0xc000090660, 0x0, 0x0)
	/go/src/app/cmd/site-controller/controller.go:129 +0x36c
main.main()
	/go/src/app/cmd/site-controller/main.go:55 +0x117
panic: assignment to entry in nil map [recovered]
	panic: assignment to entry in nil map

goroutine 1 [running]:
k8s.io/apimachinery/pkg/util/runtime.HandleCrash(0x0, 0x0, 0x0)
	/go/pkg/mod/k8s.io/apimachinery@v0.17.0/pkg/util/runtime/runtime.go:55 +0x105
panic(0x13225a0, 0x16466a0)
	/usr/local/go/src/runtime/panic.go:679 +0x1b2
github.com/skupperproject/skupper/client.(*VanClient).RouterUpdateVersionInNamespace(0xc0002a2000, 0x168e720, 0xc000048230, 0x0, 0xc000392900, 0xa, 0xc0000c3dd0, 0xc0000a8000, 0x26)
	/go/src/app/client/router_update.go:140 +0x3a10
main.(*SiteController).updateChecks(0xc000205720)
	/go/src/app/cmd/site-controller/controller.go:444 +0xff
main.(*SiteController).Run(0xc000205720, 0xc000090660, 0x0, 0x0)
	/go/src/app/cmd/site-controller/controller.go:129 +0x36c
main.main()
	/go/src/app/cmd/site-controller/main.go:55 +0x117
	```

**Comment by grs** [comment]
Sorry once again! The workaround there would be to add an annotation to the skupper-router-console service before running the update. However I will get a fix out for that also.

**Comment by grs** [comment]
Should be fixed in 0.5.3. Thank you for taking the time to report these issues and for your patience!

**Comment by Kampe** [comment]
Absolutely!
